Add tests for SocketModeReceiver (#750)

filmaj opened the pull request:
###  Summary

Saw there was a low priority "add tests for this untested thing" task (#750) so I thought I'd give it a shot! This is a first pass at adding tests for the `SocketModeReceiver` module.

I did have trouble replicating the way the e.g. [`ExpressReceiver` test create a new `FakeServer` instance in its `beforeEach` setup method](https://github.com/slackapi/bolt-js/blob/main/src/receivers/ExpressReceiver.spec.ts#L21-L24). When I tried to use the same approach, I would get this TypeScript error:

```
src/receivers/SocketModeReceiver.spec.ts:49:35 - error TS2339: Property 'fakeCreateServer' does not exist on type 'Suite'.

49         withHttpCreateServer(this.fakeCreateServer),
                                     ~~~~~~~~~~~~~~~~
```

By moving the FakeServer instantiation to within the test, that avoided the issue - though I'd like to keep the test style consistent across test files if possible! Not sure why copying the approach from `ExpressReceiver` didn't work for me 🤔 . From what I can tell, the type of `this` within a `beforeEach` block is `Context` whereas the type of `this` within an `it` block is `Suite` - so I'm not sure how it works in `ExpressReceiver` ❓ 

### Requirements (place an `x` in each `[ ]`)

* [x] I've read and understood the [Contributing Guidelines](https://github.com/slackapi/bolt/blob/main/.github/contributing.md) and have done my best effort to follow them.
* [x] I've read and agree to the [Code of Conduct](https://slackhq.github.io/code-of-conduct).

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/slackapi/bolt-js?pullRequest=1012) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/slackapi/bolt-js?pullRequest=1012) before we can accept your contribution.<br/><hr/>**Filip Maj** seems not to be a GitHub user. You need a GitHub account to be able to sign the CLA. If you have already a GitHub account, please [add the email address used for this commit to your account](https://help.github.com/articles/why-are-my-commits-linked-to-the-wrong-user/#commits-are-not-linked-to-any-user).<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/slackapi/bolt-js?pullRequest=1012) it.</sub>

filmaj commented on src/receivers/SocketModeReceiver.spec.ts:
```diff
@@ -0,0 +1,288 @@
+  };
+}
+
+// Fakes
```
All of these helper methods / classes down at the bottom were shamelessly copy-pasted from `ExpressReceiver.spec.ts`. If it would be desirable to factor this out and not copy it around, let me know, happy to do it!

seratch replied:
I think duplicating these few lines of code is okay for now. If we add more functionalities, we can revisit!

filmaj commented on src/receivers/SocketModeReceiver.spec.ts:
```diff
@@ -0,0 +1,288 @@
+        },
+      });
+      assert.isNotNull(receiver);
+      receiver.installer = installProviderStub as unknown as InstallProvider;
```
I must admit I don't have a ton of experience with TypeScript, and I feel shame for this `as unknown as whatever` type casting circus I've got going on here; if anyone knows of a cleaner way to do this, I would love to learn. FWIW, the typescript compiler told me to do this. So I blame the compiler 😇

seratch commented on src/receivers/SocketModeReceiver.spec.ts:
```diff
@@ -0,0 +1,288 @@
+/* eslint-disable @typescript-eslint/no-unsafe-call, @typescript-eslint/naming-convention */
```
Adding this test suite is a great start! 

In addition to the tests in this PR, we can add a test case to run `await receiver#stop()`. I know the test won't be a crucial test asset, but it slightly improves the code coverage. On the other hand, we cannot run `receiver#start()` as it blocks the test execution and the test results in a timeout error.

Also, mainly for your information, let me mention a few other things. As these are nice-to-haves in future development, you don't need to have anything further in this PR.

* Currently, we don't have any integration tests using both `App` and any of the built-in receivers. [App.spec.ts](https://github.com/slackapi/bolt-js/blob/main/src/App.spec.ts) uses only `FakeReceiver`, which does not do anything. We may want to have different ways to ensure the combination of App and a receiver works as we expect.

* (This may be more like an improvement on `@slack/socket-mode` side) In the future, we may want to have some tests verifying WebSocket communications (incl. auto-reconnection) and payload handling code. Currently, we don't have any tests for this, even on the `@slack/socket-mode` library side. Having some tests for the pattern will make future development / debugging reported issues.
  * For your reference, we have some tests and mock WS servers in Python/Java projects:
    * https://github.com/slackapi/bolt-python/blob/main/tests/adapter_tests/socket_mode/mock_socket_mode_server.py
    * https://github.com/slackapi/python-slack-sdk/blob/main/tests/slack_sdk/socket_mode/mock_socket_mode_server.py
    * https://github.com/slackapi/java-slack-sdk/blob/main/bolt-socket-mode/src/test/java/test_locally/SocketModeAppTest.java

I hope these are helpful to you (and anyone who read this comment!)

seratch reviewed (approved):
Looks great to me; You're so quick! 

I left a few comments. If you also think we should have one more test case for the `stop` method, update the test before merging this PR.

Note that we usually merge PRs with squashed one commit (as long as there is no reason to have multiple commits). You can do either force-push with a squashed commit or using the "Squash and merge" button in the GitHub web UI.

misscoded reviewed (approved):
LGTM! Always happy to have more tests 🙌🏼 

> From what I can tell, the type of `this` within a `beforeEach` block is `Context` whereas the type of `this` within an it block is `Suite` - so I'm not sure how it works in ExpressReceiver ❓

I believe the cause of it grabbing onto `Suite` instead of `Context` is the use of arrow functions (ie, no bound `this`). Switching them out for `function` (which is used in the `ExpressReceiver` tests) is one way to address this and should make it possible to incorporate the `beforeEach` block to contain the more repetitive bits.

filmaj commented:
Thanks for the reviews @misscoded @seratch !

> I believe the cause of it grabbing onto Suite instead of Context is the use of arrow functions (ie, no bound this). Switching them out for function (which is used in the ExpressReceiver tests) is one way to address this and should make it possible to incorporate the beforeEach block to contain the more repetitive bits.

Thanks for explaining that @misscoded, this is today's edition of "Today I Learned" 😊 . That indeed addressed the issue and I think it looks a bit cleaner now.

> If you also think we should have one more test case for the stop method, update the test before merging this PR.

Good call @seratch, I've added a very basic test for both `start` and `stop` methods (simply checks that the appropriate `SocketModeClient` underlying method is invoked).

> Also, mainly for your information, let me mention a few other things [regarding integration testing in the other client SDKs]...

This is super handy @seratch, thanks for pointing it out. I will take some time to read through these and have a think about how bolt-js could follow a similar path to ensure the same level of integration testing coverage.

> Note that we usually merge PRs with squashed one commit (as long as there is no reason to have multiple commits). You can do either force-push with a squashed commit or using the "Squash and merge" button in the GitHub web UI.

I've rebased this branch with the latest from `main` and squashed everything down to a single commit.

Assuming the latest changes are approved by you all, I'm not sure if it is OK to merge my own PR or not on the team (still figuring out The Way We Do Things) but even if that is the case, feel free to merge it in if you think it is acceptable.

This was fun 😄

seratch commented:
@filmaj Thanks for having the tests for both start and stop methods. Indeed, we can easily have those using the stub objects for the underlying `SocketModeClient` 👍 

>Assuming the latest changes are approved by you all, I'm not sure if it is OK to merge my own PR or not on the team (still figuring out The Way We Do Things) but even if that is the case, feel free to merge it in if you think it is acceptable.

You can merge it on your own now but I would like to have the honor of merging your first-ever pull request for this project 😄 Great work 🎉
